### ISSUE TITLE
Automatic updates

### ISSUE THREAD

**Issue by wordlessly** [TeamNewPipe/NewPipe#2361]
I use many apps that do an update check at startup and ask if one wants to automatically download and install the newest update. You press yes and the rest doesn't require any further user interaction.

Some even insist on updating the app. You cannot continue otherwise.

I am using NewPipe by downloading the latest from "Github releases". Sadly it doesn't auto update. Usually I check manually for a new version when Youtube changed something and NewPipe is consequently broken.

Automatic updates would be really helpful.

**Comment by Crowned-Eagle** [comment]
Install F-droid and apk release of NewPipe from there. F-droid App Store provide notifications of New App Updates and if also provides an option to download autpmatically the new version of app and notify user to update when download is complete.

You can install F-droid from https://www.f-droid.org/

Though F-droid version is one point version behind the main release. (this may be due the fact that F-Droid is OpenSource Repository and and promote uses of Non-proprietary algorithms so this may take some time to verify before listing new version to repository. Still not sure this is correct or not)

**Comment by wordlessly** [comment]
I did use NewPipe via F-Droid a couple of months back. Actually it was the sole reason why I installed F-Droid. But then things broke because of changes at Youtube's end and the only option was to install the latest Github release. F-Droid didn't offer the newest release where the problem was fixed.

If NewPipe is broken and a new release will fix it, I will get that release. I cannot wait for days or weeks for F-Droid the offer that update. Actually I could wait - but my kids can not. I use NewPipe to download audiobooks for my children for longer drives in the car. I cannot risk the anarchy and mayhem if they don't get this because F-Droid is bumbling with the updates ;)

**Comment by Crowned-Eagle** [comment]
True.. F-droid release speed too slow..

**Comment by wordlessly** [comment]
I love NewPipe and you are doing such a wonderful job. 

Automatic updates would really help to get the latest best version of your work into the hands of users. I was just wondering what the reason is, that there are no automatic updates with the Github releases? I know many apps not on the Google Play Store who do automatic (no interaction by users needed). 

Is there a technical reason why NewPipe is not doing this? Why can't the app just check for new releases hosted on Github and if successful just download and install them?

**Comment by krtkush** [comment]
Are you sure are not getting the update notifications on the Github APKs?

This feature has been there since the beginning of the year - 

https://github.com/TeamNewPipe/NewPipe/pull/1608

**Comment by wordlessly** [comment]
Actually I got that notification the first time this week :)
The first time I clicked it, it switched Chrome but then nothing happened. I had many tabs in Chrome open so maybe I missed it.

So I had to de-activate notifaications in NewPipe, shut it down, re-activate notifications again, shut it down, then re-start NewPipe. That's when I saw the notification again. Before clicking it I closed all tabs in Chrome and shut it down.
Then I clicked the notification and this time Chrome opened and showed me a "do you want to download Newpipe.xyz.apk? Now it worked.

It didn't show the download-request the first time in Chrome. Maybe I missed it. Still, this update mechanism seems flimsy.

As I said before some apps that are not listed in the Google Store manage to do that update.apk download by themselfes...without notifications or Chrome.

The most foolproof mechanism I have seen so far is an app that checks for updates at startup. It doesn't even allow you to use the app if the update check fails or if you decline to update. You have to update. This is okay because the app I am talking about needs to be connected to the internet all the time - it doesn't make sense to use it otherwise.
If you click update it will download the new apk and do all the rest - you just have to confirm the install at the end.

I don't say NewPipe should get the exact same update mechanism. You can use NewPipe offline so enforced updates makes no sense. Still, their mechanism is as foolproof as it gets.

**Comment by krtkush** [comment]
> So I had to de-activate notifaications in NewPipe, shut it down, re-activate notifications again, shut it down, then re-start NewPipe. That's when I saw the notification again.

You don't really have to do that. Just kill the app form recents and restart - you should get the notification again.

It does not make sense to force update the app but the process surely can be made more intuitive. Thanks for the feedback. I'll work on improving it.

**Comment by bew** [comment]
Why close @Stypox ?

**Comment by Stypox** [comment]
Because automatic updates have now been implemented and work correctly. Am I missing something?

**Comment by helroz** [comment]
> Because automatic updates have now been implemented and work correctly. Am I missing something?

Hello, same problem here, Newpipe check new version and send notification, when you click on it, chrome open and close a new tab without any download.

**Comment by eGit** [comment]
Yeah. This is going on for over a year I think. What I do is, I just click on the update notification, knowing it won't work in Chrome. Then I shut down Newpipe and start it again. Then it will show the notification again and the second time I try this in Chrome it always works.

I don't think you should rely on the browser to download your updates for you. Many apps just download the update themselves and it always works like a charm.

**Comment by helroz** [comment]
For me, the only way is manual download when i see the notification.

Tried the notification three times in a row and always the same result, a window that opens and closes with nothing.

**Comment by eGit** [comment]
Did you close Chrome by any chance between tries? Wouldn't do that.

Otherwise I have no clue how to fix your problem. My little workaround probably doesn't work for everybody.

**Comment by wb9688** [comment]
The way to fix that is to either use a proper browser, or for us to link the GitHub releases page instead.

**Comment by eGit** [comment]
What technical limitation is there for Newpipe to download the updates itself? I have seen other apps do this and it works perfectly.

Why can other apps do it but Newpipe cannot?

**Comment by wb9688** [comment]
@eGit: No technical limitations, but we would prefer you to use F-Droid instead, and want to improve the notification to link to the F-Droid app when you've installed it from F-Droid, or to the GitHub releases page if that's where you downloaded it from.

**Comment by eGit** [comment]
I was using F-Droid just for Newpipe. But it was always lagging behind with updates and updates are critical for Newpipe whenever Youtube is a dick, which is basically ALL THE FUCKING TIME.

Don't really see the point in F-Droid. In principle it is good but in practice most people need Google Play Store. I would use F-Droid if I could drop Google. Otherwise it doesn't make much sense to run app stores in parallel.

**Comment by eGit** [comment]
Why not make a poll and see what kind of update mechanism users want? I bet most users want automatic updates done by Newpipe itself. You start Newpipe, it downloads the update in the background and then sends you a notification that a new update is available and if you click the notification it will be applied. Optionally Newpipe could show the user an update/restart dialog whenever a video throws an error. Perfect time to tell the user that an update is in "the pipe".

**Comment by helroz** [comment]
I don't close chrome between tries ,only close newpipe.
A pool is a good idea, fdroid is not for me, i have updates with a long long time, last time used it, no update, see it 48h after only.
(Sorry for my english^^).

**Comment by opusforlife2** [comment]
> But it was always lagging behind with updates

Newpipe has its _own_ F-Droid repo now, though. It'll be updated at the same time as the Github version.

**Comment by eGit** [comment]
Yep. F-Droid doesn't work very well with Newpipe and I tried it several times on different devices. Just installed it for Newpipe.

As a side note: many Newpipe users are children as young as three. They cannot deal with the Update Notification especially since it doesn't work. Then they come running and yell, Dad, Dad, Newpipe is broken again. So I do the update workaround or Google the Newpipe release page and pray to god that I find the official repo and don't install malware.

So yeah. Please make it as easy as possible. Let Newpipe update itself with as little user interaction as possible. People who want to use F-Droid can still use it. I don't however get the obsession of Newpipe with F-Droid. Newpipe is such an AMAZING app and the update experience ruins all of it.

**Comment by krtkush** [comment]
I use Firefox and the download via notification works seamlessly. Give that a try?

**Comment by eGit** [comment]
Yeah sure. Install F-Droid and install Firefox and do your updates manually just to use a thin layer over Youtube :(

Guys, at some point the benefit of Newpipe over Youtube is void because it is more trouble then worth. Newpipe needs updates all the time in order to keep working. It's not updated every 6 months! Apps that update frequently shouldn't do this manual browser bull.

Automatic background updates are the only way!

And yes, I installed Firefox before. It didn't work either. Manual updates via Browser probably works for 70% of the people but not for 30% for whatever reason. Android Version, Browser version, whatever. And 30% is just horrible especially when you factor in that this is a manual update. Horrible experience.

**Comment by helroz** [comment]
I have tried with firefox, 1 trie, 1 success.

**Comment by helroz** [comment]
+1 for eGit, updates by newpipe app is the best way.

I don't reinstall fdroid, i use miui and need to protect it with the taskkiller for working good. 
Use firefox.... I have all in chrome, not firefox.

For me manual update is the only way actually.

**Comment by eGit** [comment]
This is so embarrasing. I gave F-Droid a second chance. Maybe Newpipe or F-Droid fixed the issue where the version of Newpipe in their store was not totally out of date.

Just started F-Droid. They currently offer version 20.1, released at October the 17th.

There have been 20.2, 20.3, 20.4, 20.5 and 20.6 (the current version at the moment).

NOTHING works in 20.1. Youtube changed everything by now. 20.1 is JUST BROKEN!

Why even offer Newpipe in F-Droid when it is hopelessly out of date! Either a new version appears on Github AND F-Droid on the SAME DAY or it is just a broken way for deployment!

Why have a genius piece of software like Newpipe and then mess the deployment big time!

The current update mechanism via Browser is STILL HOPELESSLY UNRELIABLE! I still have to repeat it 2-3 times until it works. Close Newpipe, start it again, see the Notification, click on it, open the browser, download (which does NOT happen every time), install. I have to repeat this grueling process 2-3 times! And you update Newpipe every other week. This is just a BIG FAT JOKE!

Automatic updates! Do it Now!

**Comment by ynamsa** [comment]
I have the same problem. 
I would really like to request the automatic update.

**Comment by bew** [comment]
Hello,

I'd like to recommend this app for non technical people, who does not now how to install apps from github, or even how to update the app if the current "easy method" does not work.

I've been reluctant to install the app on people's phone as a replacement of youtube for this reason.

**Comment by dm17** [comment]
How do you trigger the update manually? A few times I've see an "update is available" pop-up... Then I often lose the notification and not sure how long I must wait for it to notify again.
Thanks.

**Comment by eGit** [comment]
> How do you trigger the update manually? A few times I've see an "update is available" pop-up... Then I often lose the notification and not sure how long I must wait for it to notify again.
> Thanks.

No point. This is a long standing issue. The devs see no problem although people are complaining all the time. I gave up using Newpipe and switched back to Youtube: SADLY, because their Ads are super annoying as well. But all in all it works. The broken Newpipe update mechanism forces me to do this.

Newpipe: one of the best apps out there, with a real demand and all for nothing because they cannot fix their update procedure. Pornhub is the opposite. The app sucks but the update mechanism is the best I have ever seen. Maybe they should team up.

**Comment by dm17** [comment]
@eGit the fact that you know anything about the pornhub app is concerning. Perhaps the NewPipe devs have an easy way to trigger updates. It is no wonder you perceive NewPipe as such a massive inconvenience now that you're accustomed to pleasuring yourself to videos of drug addled runaways, many of whom have already committed suicide... Perhaps it is something to consider when trying to optimize life for pleasure and convenience.

**Comment by eGit** [comment]
> @eGit the fact that you're talking openly about using porno applications is a big detour in this thread. Perhaps the NewPipe devs have an easy way to trigger updates. It is no wonder you perceive NewPipe as such a massive inconvenience now that you're accustomed to pleasuring yourself to videos of drug addled runaways, many of whom have already committed suicide... Perhaps it is something to consider when trying to optimize life for pleasure and convenience.

The PH app was pointed out as having the most convenient, most fool-proof update mechanism in existence. I don't use it but whatever. No need to justify myself to an incel.
